provisionerd/runner.go

```diff
@@ -0,0 +1,872 @@
+	// to this is if something calls fail() on the runner; either something external, like the server getting closed,
+	// or the heartbeat goroutine timing out after attempting to gracefully cancel.  If something calls fail(), then
+	// the failure is sent on that goroutine on the context passed into fail(), and it marks okToSend false to signal
+	// us here that this function should not also send a terminal message.
+	go r.run()
+	go r.heartbeat()
+	for r.failedJob == nil && r.completedJob == nil {
+		r.cond.Wait()
+	}
+	if !r.okToSend {
+		// nothing else to do.
+		return
+	}
+	if r.failedJob != nil {
+		r.logger.Debug(r.forceStopContext, "sending FailedJob")
+		err := r.sender.failJob(r.forceStopContext, r.failedJob)
+		if err != nil {
+			r.logger.Error(r.forceStopContext, "send FailJob", slog.Error(err))
+		}
+		r.logger.Info(r.forceStopContext, "sent FailedJob")
+	} else {
+		r.logger.Debug(r.forceStopContext, "sending CompletedJob")
+		err := r.sender.completeJob(r.forceStopContext, r.completedJob)
+		if err != nil {
+			r.logger.Error(r.forceStopContext, "send CompletedJob", slog.Error(err))
+		}
+		r.logger.Info(r.forceStopContext, "sent CompletedJob")
+	}
+	close(r.done)
+	r.okToSend = false
+}
+
+// cancel initiates a cancel on the job, but allows it to keep running to do so gracefully.  Read from isDone() to
+// be notified when the job completes.
+func (r *runner) cancel() {
+	r.cancelFunc()
+}
+
+func (r *runner) isDone() <-chan any {
+	return r.done
+}
+
+// fail immediately halts updates and, if the job is not complete sends FailJob to the coder server.  Running goroutines
+// are canceled but complete asynchronously (although they are prevented from further updating the job to the coder
+// server).  The provided context sets how long to keep trying to send the FailJob.
+func (r *runner) fail(ctx context.Context, f *proto.FailedJob) error {
+	f.JobId = r.job.JobId
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	if !r.okToSend {
+		return nil // already done
+	}
+	r.cancel()
+	if r.failedJob == nil {
+		r.failedJob = f
+		r.cond.Signal()
+	}
+	// here we keep the original failed reason if there already was one, but we hadn't yet sent it.  It is likely more
+	// informative of the job failing due to some problem running it, whereas this function is used to externally
+	// force a fail.
+	err := r.sender.failJob(ctx, r.failedJob)
+	r.okToSend = false
+	r.forceStopFunc()
+	close(r.done)
+	return err
+}
+
+// setComplete is an internal function to set the job to completed.  This does not send the completedJob.
+func (r *runner) setComplete(c *proto.CompletedJob) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	if r.completedJob == nil {
+		r.completedJob = c
+		r.cond.Signal()
+	}
+}
+
+// setFail is an internal function to set the job to failed.  This does not send the failedJob.
+func (r *runner) setFail(f *proto.FailedJob) {
```

Comment:
nit: The difference between `fail` and `setFail` to me isn't immediately obvious. I don't have any better recommendations, but I think it could be confusing.

Reply:
I couldn't think of anything better either!  So, I wrote some comments...

Reply:
Reasonable